### ISSUE TITLE
[TASK] Final report - Future Work

### ISSUE THREAD

**Issue by Jarrus00** [WSU-CPTS415-ParquetParkour/Amazon-CoPurchasing#40]
_Background_:
There have been a number of things learned during the course of this project and alternate options which we did not have the bandwidth to explore.  These should be documented and included at the end of the report, prior to Contributions.  This is entirely optional, but would be nice to have.

_Problem_:
What was learned during this project, separate from the course teachings?  What could have we have explored with more bandwidth?

_Success Criteria_:
1. Areas which could be further improved and suggestions on how to do so are documented.
2. References are cited and included in a subsection at the end, if needed.

**Comment by Jarrus00** [comment]
While we sought to optimize processes and algorithms throughout development, there were a number of factors discovered which could lead to a more expedient and robust operation.  Firstly, while we implemented a degree of parallel processing to prepare the source data for ingestion to the database, there were bottlenecks in the process where data is being merged and undergoing feature extraction which could be addressed by refactoring to leverage a stream processing approach via Apache Kafka and/or Apache Spark.  Making use of a CI/CD or data workflow tool to construct a data pipeline while iterating on the parser and ingestion process would help to minimize the number of manual steps required to process the source data and populate the database.  With expanding the environment, making use of multiple machines with SSDs in a RAID 0 configuration or multiple containers in a cloud environment to distribute the database in a shard configuration would help improve the processing time for aggregate functions as this was a notable throughput restriction in our experiments.  Another constraint in this project was the use of Neo4j with a Community license which presents a reduced feature set, and more robust processes are expected to be possible with the Enterprise license, or even using the Aura instance had the dataset been notably smaller due to restrictions on database size.

The sample set provided to the collaborative filtering algorithm had a high probability of being sparse (see below figure), frequently resulting in insufficient review values to rank recommendations by and revisiting this via dimensional reduction (i.e.: principal component analysis) to minimize the volume of products.  Implementing a compressed sparse matrix format prior to data transport would be helpful here as well to shrink the footprint of the data being analyzed.

![review_count_densities](https://user-images.githubusercontent.com/71807740/206925262-6e48fd19-b371-4087-81d2-144893156f43.png)

In the UI, we restricted user input to numeric comparisons only to control scope and ensure delivery timelines.  This reduces the value presented by the application and opening up functionality to include text searches would be a benefit to the users.  Further expansion of the UI to include an ingestion interface and methods to save results would also improve the user experience, allowing for the comparison of outcomes.
